### ISSUE TITLE
creating record with a json attribute type fails

### ISSUE THREAD

**Issue by orthodoc** [balderdashy/sails-postgresql#155]
model User.js
```
module.exports = {
  attributes: {
    username: { type: 'string', unique: true },
    password: { type: 'string' },
    ....
    socialProfiles: { type: 'json', defaultsTo: {} },
  },
}
```
Returns the following error:
```
{ code: 'E_INTERNAL_SERVER_ERROR',
  message: 'Something bad happened on the server',
  data: 
   Error (E_UNKNOWN) :: Encountered an unexpected error
   error: column user.socialProfiles.facebook.id does not exist
       at Connection.parseE (/home/bdb/sails/diagnox/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:534:11)
       at Connection.parseMessage (/home/bdb/sails/diagnox/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:361:17)
       at Socket.<anonymous> (/home/bdb/sails/diagnox/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:105:22)
       at Socket.emit (events.js:107:17)
       at readableAddChunk (_stream_readable.js:163:16)
       at Socket.Readable.push (_stream_readable.js:126:10)
       at TCP.onread (net.js:529:20)
   
   Details:  error: column user.socialProfiles.facebook.id does not exist
    }
```
When I set the model connection to `memory` instead of `postgresql`, the record is successfully executed

Referring to ghaiklor/generator-sails-rest-api#44

**Comment by dogrocker** [comment]
Did you try
```
module.exports = {
 schema: true,

  attributes: {

  username: { type: 'string', unique: true },
  password: { type: 'string' },
  ....
  socialProfiles: { type: 'json', defaultsTo: {} },
  }
}
```

**Comment by orthodoc** [comment]
It doesn't work with the `schema: true` option as well.

**Comment by orthodoc** [comment]
This is the snippet of code from config/passport.js where it is failing.
```
User                                                                                                                                                                                                                                                                   
  .findOrCreate(criteria, model)                                                                                                                       
  .exec(function (error, user) {                                                                                                                       
    if (error) return next(error, false, {});                                                                                                          
    if (!user) return next(null, false, {                                                                                                              
       code: 'E_AUTH_FAILED',                                                                                                                           
       message: [profile.provider.charAt(0).toUpperCase(), profile.provider.slice(1), ' auth failed'].join('')                                          
     });                                                                                                                                                
                                                                                                                                                           
  return next(null, user, {});                                                                                                                       
  });
```
`console.log(criteria)` returns the following `{ 'socialProfiles.facebook.id': '101xxxxxxxxxxxx15' }`

It appears to me that the `findOrCreate` method is unable to execute when the json attribute is empty.

**Comment by dogrocker** [comment]
may be id type will be bigint? because facebook id is long digit can't use int to handle that.

**Comment by orthodoc** [comment]
@dogrocker the `socialProfiles` attribute is of type `json` and it is empty. In other words, `socialProfiles.facebook` and `socialProfiles.facebook.id` are not set to be of any type, and therefore it shouldn't matter if the id is a `bigint` or not.

**Comment by dmarcelino** [comment]
@orthodoc, I haven't used postgres in ages, can you generally query by json attribute like `socialProfiles.facebook.id` even if you have empty `socialProfiles` records?

`findOrCreate()` essentially does a `find()` and if no records exist then proceeds to do a `create()`. Can you try doing the 2 operations separately in order to understand which operation causes the issue?

**Comment by orthodoc** [comment]
Thanks @dmarcelino for your suggestions. 

I have tried the same and reported it [generator-sails-rest-api/issues/44](https://github.com/ghaiklor/generator-sails-rest-api/issues/44#issuecomment-96769759).

**Comment by dmarcelino** [comment]
> When I replace the `findOrCreate` method with `find` I get exactly the same error: 
`error: column user.socialProfiles.facebook.id does not exist`

As I suspected, your issue is the `find` because waterline does not support the dot `.` notation nor it supports querying inside a `json`. The way to go about querying nested associations would be to use associations ([docs](https://github.com/balderdashy/waterline-docs/blob/master/associations.md)).

Given that you probably want to stick with the `json` you can also try to find a SQL query that would suite your needs and use the `.query()` methods ([docs](https://github.com/balderdashy/waterline-docs/blob/master/query-methods.md#query-query-data-callback-)).

**Comment by ghaiklor** [comment]
@dmarcelino any chance to implement dot.notation for postgresql adapter ?

**Comment by dmarcelino** [comment]
I doubt there is much appetite for that since it's not supported by the waterline API... Regarding which there is a discussion about adding better support for `embedded` documents in balderdashy/waterline#658.

**Comment by ghaiklor** [comment]
@dmarcelino I know that dot.notation is working for mongo adapter. I'm always use sails-mongo and find criteria like `socialProfiles.facebook.id` works great. So should be the way do it for postgresql, imho.

**Comment by devinivy** [comment]
The primary difference is that mongodb supports that functionality natively for embedded documents: http://docs.mongodb.org/manual/core/document/#dot-notation

But there's no native way to do this in postgresql of which I am aware.

**Comment by dmarcelino** [comment]
*I'm not familiar enough with this feature but I'll conjecture a bit.* I believe it works on mongo adapter more due to the [features of mongoDB](http://docs.mongodb.org/manual/core/document/#dot-notation) than due to any particular work trying to convert the dot.notation into a *mongoDB query*. Given the differences between MongoDB and PostgreSQL it may not be that easy to port it to sails-postgresql, or other adapters for that matter.

**Comment by devinivy** [comment]
I have seen people hack around this by querying on the stringified version that lives in the database.  Ex,
```node
Users.find({
    socialProfiles: {
        contains: JSON.stringify({facebook: {id: 10}})
    }
}).exec(function(err, users){});
```

But don't expect it to be as flexible as mongodb!

**Comment by ghaiklor** [comment]
@dmarcelino @devinivy I see. So dot.notation feature it's just mongodb feature :smile: 
@devinivy but your workaround doesn't guarantee that it returns unique record from db.

**Comment by orthodoc** [comment]
Just for the [record](https://github.com/ghaiklor/generator-sails-rest-api/issues/44#issuecomment-101021576). `jsonb` type allows deep nested querying in postgresql-9.4

**Comment by sailsbot** [comment]
Thanks for posting, @orthodoc.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
